Add method to save Cadet sim as python file which can generate the Cadet sim again

ronald-jaepel opened the pull request:
Also add test for the save_as_python method
Also changed a few lines in cadet.py to follow PEP 8

ronald-jaepel commented:
This has been rebased and the tests still run. What was the reason this wasn't merged again?

schmoelder commented on tests/test_save_as_python.py:
```diff
@@ -0,0 +1,64 @@
+
+    # populate "sim" variable using the generated code lines
+    for line in code_lines:
+        exec(line)
```
Tbh, I'm not entirely sure about exec either.

schmoelder replied:
Sorry for not being more constructive, it was quite a busy day. I can have another look tomorrow.

schmoelder reviewed (approved):
Almost there...

I added docstrings and type annotations.

Moreover, I also refactored the method s.t. it not only works for `Cadet` objects but also more generically also for `H5` types (where the method lives) or other derived classes.

schmoelder commented on cadet/h5.py:
```diff
@@ -520,3 +577,64 @@ def recursively_save(h5file: h5py.File, path: str, dic: Dict, func: callable) ->
+    Recursively turn a nested dictionary or addict.Dict into a list of Python code that
+    can generate the nested dictionary.
+
+    :param dictionary:
```
let's also be consistent with formatting. In CADET-Python, we use numpy style docstrings and a max line length of 88 chars.

schmoelder replied:
Also, please use numpy style documentation for consistency.

schmoelder commented:
I just talked to @jbreue16. He would like to test this with a simulation file that also contains output. :nerd_face: 

Maybe we need some check on the size of the parameters or we should add an argument to specify / limit the branches to be "imported".

Other than that, looks really good. Thanks for pushing it; I had already forgotten about it. :sweat_smile:

jbreue16 commented:
Writing the attached h5 with the output data to a python file using the added functionality, I got e.g.
`model.root.output.solution.unit_006.solution_outlet = np.array([6.52177391e-26, 6.46449692e-26, 6.40770758e-26, ...,
       4.58453836e-05, 4.54447327e-05, 4.50474730e-05], shape=(1500,))`
We should add an exception whenever that happens.

I also got lines like
`model.root.input.model.unit_000.adsorption.lin_kd = np.np.float64(1.0)`
with np.np calls ...

ronald-jaepel commented:
Thanks for the additional test-case! It helped catch two extra cases that I've been able to fix two of.

>     * long arrays are abbreviated with dots: `model.root...output.solution.unit_006.solution_outlet = np.array([6.52177391e-26, 6.46449692e-26, 6.40770758e-26, ..., 4.58453836e-05, 4.54447327e-05, 4.50474730e-05], shape=(1500,))`
>       We should add an exception whenever that happens.

Is fixed by 1416997a3074ff5cffb074f09754b8a2ccb36b8d

On my machine it threw an error for `np.array([b'EQUIDISTANT_PAR'], dtype='|S15')` during the equality assertion because it couldn't handle the dtype, so I added an extra catch for that. I've also changed the way we compare arrays to use relative tolerances instead of absolute tolerances in 9df19fe10dc4177183cd7a321273236273445f91

regarding: 
>     * additional dots after root: `model.root...input.model.unit_000.adsorption.lin_kd = np.np.float64(1.0)`

I don't observe that. Is this still present in the current version of the PR for you?

Regarding:

ronald-jaepel commented:
I can't seem to re-run the CI pipeline. I'll close this PR and re-open one to trigger the CI.

ronald-jaepel commented:
Re-opening the PR helped to run the CI. All tests pass.

jbreue16 commented:
Looks good to me, thanks Ron !
